drivers/ieee802154/ieee802154_rf2xx.c

```diff
@@ -0,0 +1,893 @@
+		}
+		k_mutex_unlock(&ctx->phy_mutex);
+	}
+}
+
+/********************
+ * Radio aux API *
+ *******************/
+static inline u8_t *get_mac(struct device *dev)
+{
+	struct rf2xx_context *ctx = dev->driver_data;
+
+	/*
+	 * The RF2XX has two 32-bit registers for the MAC address where
+	 * 40 bits of the registers are factory programmed to be unique
+	 * and the rest are to be assigned as the "company-specific" value.
+	 * 802.15.4 defines a EUI-64 64-bit address with company specific
+	 * being 24 or 36 bits with the unique value being 24 or 40 bits.
+	 */
+
+	u32_t *ptr = (u32_t *)(ctx->mac_addr);
+
+	UNALIGNED_PUT(sys_rand32_get(), ptr);
+	ptr = (u32_t *)(ctx->mac_addr + 4);
+	UNALIGNED_PUT(sys_rand32_get(), ptr);
+
+	/*
+	 * Clear bit 0 to ensure it isn't a multicast address and set
+	 * bit 1 to indicate address is locally administrered and may
+	 * not be globally unique.
+	 */
+	ctx->mac_addr[0] = (ctx->mac_addr[0] & ~0x01) | 0x02;
+
+	return ctx->mac_addr;
+}
+
+/********************
+ * Radio device API *
+ *******************/
+static enum ieee802154_hw_caps rf2xx_get_capabilities(struct device *dev)
+{
+	struct rf2xx_context *ctx = dev->driver_data;
+	int hwcaps = 0;
+
+	if (ctx->trx_model == RF2XX_TRX_MODEL_212) {
+		hwcaps = IEEE802154_HW_SUB_GHZ;
+	} else {
+		hwcaps = IEEE802154_HW_2_4_GHZ;
+	}
+
+	return IEEE802154_HW_FCS |
+	       IEEE802154_HW_PROMISC |
+	       IEEE802154_HW_FILTER |
+	       IEEE802154_HW_CSMA |
+	       IEEE802154_HW_TX_RX_ACK |
+	       hwcaps;
+}
+
+static int rf2xx_cca(struct device *dev)
+{
+	return -EBUSY;
+}
+
+static int rf2xx_set_channel(struct device *dev, u16_t channel)
+{
+	struct rf2xx_context *ctx = dev->driver_data;
+
+	if (ctx->trx_model == RF2XX_TRX_MODEL_212) {
+	/* TODO: Set RSSI base for Sub-Giga Transceiver
+	 * #define PHY_RSSI_BASE_VAL_BPSK_20             (-100)
+	 * #define PHY_RSSI_BASE_VAL_BPSK_40             (-99)
+	 * #define PHY_RSSI_BASE_VAL_OQPSK_SIN_RC_100    (-98)
+	 * #define PHY_RSSI_BASE_VAL_OQPSK_SIN_250       (-98)
+	 * #define PHY_RSSI_BASE_VAL_OQPSK_RC_250        (-97)
+	 */
+	} else {
+		if (channel < 11 || channel > 26) {
+			LOG_ERR("Unsupported channel %u", channel);
+			return -EINVAL;
+		}
+
+		u8_t reg = rf2xx_iface_reg_read(dev, PHY_CC_CCA_REG) & ~0x1f;
+		rf2xx_iface_reg_write(dev, PHY_CC_CCA_REG, reg | channel);
+	}
+
+	return 0;
+}
+
+static int rf2xx_set_txpower(struct device *dev, s16_t dbm)
+{
+	(void) dbm;
+	/* TODO: Add look-up table
+	 * Now will max power
+	 */
+	u8_t reg = rf2xx_iface_reg_read(dev, PHY_TX_PWR_REG) & ~0x0f;
+	rf2xx_iface_reg_write(dev, PHY_TX_PWR_REG, reg);
+
+	return 0;
+}
+
+static int rf2xx_set_ieee_addr(struct device *dev, bool set,
+	const u8_t *ieee_addr)
+{
+	LOG_DBG("IEEE address %02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x",
+		ieee_addr[7], ieee_addr[6], ieee_addr[5], ieee_addr[4],
+		ieee_addr[3], ieee_addr[2], ieee_addr[1], ieee_addr[0]);
+
+	if (set) {
+		const u8_t *ptr_to_reg = ieee_addr;
+		for (u8_t i = 0; i < 8; i++) {
```

Comment:
break a line between variable declaration and their actual usage

Reply:
under review

Reply:
Fixed